### DISCUSSION TITLE
On the design of the react-email API and CSS safety

### DISCUSSION THREAD

**Discussion by zxti** [resend/react-email#734] [Q&A]
Hi, thanks for making this library.

At first, I thought react-email was meant to enable safe construction of emails that display correctly in email clients, using React/JSX.

After using it, I realized it has a very porous API that admits arbitrary CSS. So you can simply add `padding` or `display` properties that (say) Yahoo email will not display correctly. Which is guaranteed to happen in our organization at scale.

What is the purpose of using react-email (over using React HTML directly), if it allows arbitrary CSS? I know that with these components I at least won't render something like a `<iframe>`, and there's a bit of boilerplate `<table>` construction in places, but it seems that the core problem in email development is really the layout and styling across myriad email clients.

Would you consider evolving the API toward greater safety and strictness, a la MJML?

**Comment by bukinoshita** [comment]
> Hi, thanks for making this library.
> 
> At first, I thought react-email was meant to enable safe construction of emails that display correctly in email clients, using React/JSX.
> 
> After using it, I realized it has a very porous API that admits arbitrary CSS. So you can simply add `padding` or `display` properties that (say) Yahoo email will not display correctly. Which is guaranteed to happen in our organization at scale.
> 
> What is the purpose of using react-email (over using React HTML directly), if it allows arbitrary CSS? I know that with these components I at least won't render something like a `<iframe>`, and there's a bit of boilerplate `<table>` construction in places, but it seems that the core problem in email development is really the layout and styling across myriad email clients.
> 
> Would you consider evolving the API toward greater safety and strictness, a la MJML?

Yes, React Email is still in progress and far from perfect. For now it helps you with basic structure and opinionated things that helps rendering in email clients.

One example is this line in the Button component, where if you're not an email expert, you don't need to know this and just use the Button component.

```
<!--[if mso]><i style="letter-spacing: ${pX}px;mso-font-width:-100%;mso-text-raise:${textRaise}" hidden>&nbsp;</i><![endif]-->
```

We do have plans on having a more strict API for styles, a linter to give you warns and errors when building emails.
